### PR TITLE
if-then -> master

### PR DESCRIPTION
**Pull request by liambaloh**
Adds an ifThen(..) function to Map as discussed in https://github.com/aimeos/map/pull/7#issuecomment-921456893. The function takes two or three callables as parameters: $condition, $then and $else. Each callable accepts the Map as a parameter. $condition should return a boolean or something which can be evaluated to a boolean. If it returns true, $then will be executed, otherwise $else will be executed (if provided).

Please let me know if there's anything which should be changed, also please feel free to modify this code in any way whatsoever, or if it's not suitable for this repository please feel free to close the PR. I'm merely sharing something I needed in my project.

### CODE REVIEW

**Comment by aimeos** [comment]
More thoughts:

- PHP uses `if()` instead of `ifThen()` so it should be obvious for all PHP developers if we name the method only `if()` and it's less to write than `ifThen()`

- It would be simpler in many cases if the condition would accept something that PHP will evaluate to a boolean or a closure so the code could be:
```
map( ['test'] )->if(
    $a == $b,
    function( $map ) { die(); },
    function( $map ) { echo 'OK'; }
);
```
or
```
map( ['test'] )->if(
    fn( $map ) => $map->empty(),
    function( $map ) { die(); },
    function( $map ) { echo 'OK'; }
);
```

The final thought is if there's any advantage at all when comparing to:
```
$map map( ['test'] );
$map->empty() ? die() : echo 'OK';
```
As you can see, it's much less code than the new `if()` method.

**Comment by liambaloh** [comment]
Went with ifThen to make it obvious from the method name what the expected order of parameters is: first one's if, second one's then. It's certainly true that PHP uses if, and not if-then in its standard language, however it's unfortunately also true that PHP has a major problem in inconsistent parameter orders in its core function, so went with more clarity than strictly necessary. Also went with ifThen(..) to make it clear that both the condition and effects are expected parameters of the same function, this is to avoid confusion with promise-based approaches which might look like
```
$promise
    ->If(fn(..) => ..)
    ->Then(fn(..) => ..)
    ->Else(fn(..) => ..)
    ->Exception(fn(..) => ..)
    ->Execute()
```

I have no opinion on closures or taking a boolean directly. It's not what I need but can add it in if desired.

As to why: It's so conditional execution can be inserted into a cascade of functions. Not to use the repository's own blurb against it, but it seems in-line with this bit:
image

Or in our case: 
Instead of 
```php
$tmp = $map
    ->filter(..)
$tmp->empty() ? die() : echo 'OK';
$value = $tmp
    ->map(..)
    ->fold(..)
```
to do
```php
$value = $map
    ->filter(..)
    ->if(
        fn(..) => $map->Empty(),
        fn(..) => die(),
        fn(..) => echo 'OK')
    ->map(..)
    ->fold(..)
```

Please let me know your thoughts on this. I have no objections to doing it in the way you propose, just thought I'd first say why I made the decisions I made.

**Comment by aimeos** [comment]
Think, it's OK if we add an `if()` method that behaves like this:
> ```
> $value = $map
>     ->filter(..)
>     ->if(
>         fn(..) => $map->Empty(),
>         fn(..) => die(),
>         fn(..) => echo 'OK')
>     ->map(..)
>     ->fold(..)
> ```
Futhermore, still think that `if()` if preferrable over `ifThen()` because the parameter order inconsistencies in PHP are mainly in the array and string methods but `if` statements work the same in all programming languges. Also, the Map class will never be promise based because it makes things much more complicated (as you can see in CakePHP).

**Comment by liambaloh** [comment]
Alright, that's fair: That has been implemented in 07f46aa5312002177c3eabc478a79dd393eb5b06

If you'd like to test before merging, please execute any of these demo scripts:
```php
Map::from( ['a' => 1, 'b' => 0] )->if(
    fn(Map &$map) => $map->has('a'),
    function(Map &$_) {echo "then";},
    function(Map &$_) {echo "else";}
);
//then

Map::from( ['a' => 1, 'b' => 0] )->if(
    fn(Map &$map) => $map->has('c'),
    function(Map &$_) {echo "then";},
    function(Map &$_) {echo "else";}
);
//else

Map::from( ['a' => 1, 'b' => 0] )->if(
    fn(Map &$map) => $map->has('c'),
    function(Map &$_) {echo "then";}
);
// (no output)
```

File: src/Map.php [hunk 1]
```diff
@@ -3727,4 +3727,34 @@ protected function visit( iterable $entries, array &$result, int $level, ?\Closu
...
+	 * @param callable $then Function with (Map) parameter and returns void
+	 * @param callable|null $else Optional function with (Map) parameter and returns void
+	 * @return self Same map for fluid interface
+	 */
+	public function if(callable $condition, callable $then, callable $else = null): self
+	{
+		if ($condition($this)) {
+			$then($this);
+		} else {
+			if ($else != null) {
```
Comment (aimeos) [inline src/Map.php]: You can combine this to one `elseif`:
```
} elseif( $else ) {
```

File: src/Map.php [hunk 2]
```diff
@@ -3727,4 +3727,34 @@ protected function visit( iterable $entries, array &$result, int $level, ?\Closu
...
+	 *              fn(Map &$Map) => $Map->isEmpty(),
+	 *              fn(Map &$_) => die("unknown page: " . $pageId))
+	 *          ->first();
+	 *
+	 * @param callable $condition Function with (Map) parameter and returns bool
+	 * @param callable $then Function with (Map) parameter and returns void
+	 * @param callable|null $else Optional function with (Map) parameter and returns void
+	 * @return self Same map for fluid interface
+	 */
+	public function if(callable $condition, callable $then, callable $else = null): self
```
Comment (aimeos) [inline src/Map.php]: Can you please remove the first `callable` type hint so it can be also something that evaluates to a boolean?

File: src/Map.php [hunk 3]
```diff
@@ -3727,4 +3727,34 @@ protected function visit( iterable $entries, array &$result, int $level, ?\Closu
...
+
+	/**
+	 * Executes $callback if the Map is empty
+	 *
+	 * Examples: (Returns match or dies if doesn't exist)
+	 *  $page = $this->pages
+	 *          ->filter(fn(Page $page, $_) => $page->Id == $pageId)
+	 *          ->if(
+	 *              fn(Map &$Map) => $Map->isEmpty(),
+	 *              fn(Map &$_) => die("unknown page: " . $pageId))
```
Comment (aimeos) [inline src/Map.php]: Please change `&$Map` to `$map` only because objects are passed by reference by default

**Comment by aimeos** [comment]
> If you'd like to test before merging, please execute any of these demo scripts:

We need automated tests. Can you please add them to the ./tests/MapTest.php file?

File: src/Map.php (hunk 1 above)
Comment (liambaloh) [reply src/Map.php]: Resolved in 7b6ad457a2dbdf619d0e3d91ea67f009ff626a4d

File: src/Map.php (hunk 2 above)
Comment (liambaloh) [reply src/Map.php]: Resolved in 7b6ad457a2dbdf619d0e3d91ea67f009ff626a4d

File: src/Map.php (hunk 3 above)
Comment (liambaloh) [reply src/Map.php]: Resolved in 7b6ad457a2dbdf619d0e3d91ea67f009ff626a4d (also in Readme.md)

**Comment by liambaloh** [comment]
Code issues resolved and tests added

File: README.md [hunk 4]
```diff
@@ -1788,6 +1789,43 @@ Map::from( ['a' => 'X', 'b' => 'Y'] )->has( 'X' );
 ```
 
 
+### if()
+
+Executes callables $then and $else depending on callable $condition.
+
+```php
+public function if( Callable $condition, Callable $then, $Callable $else = null ) : bool
```
Comment (aimeos) [inline README.md]: The type hint for the first parameter must be removed too

File: README.md [hunk 5]
```diff
@@ -1788,6 +1789,43 @@ Map::from( ['a' => 'X', 'b' => 'Y'] )->has( 'X' );
...
+```
+* @param callable $condition Function with (Map) parameter and returns bool
```
Comment (aimeos) [inline README.md]: Type hint must be `callable|bool`. Also, the description should state that it must evaluate to bool, not return bool

File: src/Map.php (hunk 3 above)
Comment (aimeos) [reply src/Map.php]: Not yet, it's still in the method doc header

File: src/Map.php [hunk 6]
```diff
@@ -3727,4 +3727,37 @@ protected function visit( iterable $entries, array &$result, int $level, ?\Closu
...
+		$conditionResult = $condition;
+		if($condition instanceof \Closure){
+			$conditionResult = $condition($this);
+		}
+
+		if ($conditionResult) {
+			$then($this);
+		} elseif($else != null) {
+			$else($this);
+		}
```
Comment (aimeos) [inline src/Map.php]: You can shorten this to:
```
if( $condition instanceof \Closure ? $condition( $this ) : $condition ) {
	$then( $this );
} else {
	$else( $this );
}
```

File: tests/MapTest.php [hunk 7]
```diff
@@ -2659,6 +2659,33 @@ public function testZip()
...
+	}
+
+	public function testIfBlankElse()
+	{
+		Map::from( ['a' => 1, 'b' => 0] )->if(
+			fn(Map $map) => $map->has('c'),
+			function(Map $_) { $this->assertTrue(false); }
+		);
+		$this->assertTrue(true);
+	}
```
Comment (aimeos) [inline tests/MapTest.php]: Can you please move this up between testHas*() and testIn*()?

File: README.md (hunk 4 above)
Comment (liambaloh) [reply README.md]: fixed in c329afdda114015b900470c1ef73edc6439fc1a1

File: README.md (hunk 5 above)
Comment (liambaloh) [reply README.md]: fixed in c329afdda114015b900470c1ef73edc6439fc1a1

File: src/Map.php (hunk 6 above)
Comment (liambaloh) [reply src/Map.php]: fixed in c329afdda114015b900470c1ef73edc6439fc1a1

File: tests/MapTest.php (hunk 7 above)
Comment (liambaloh) [reply tests/MapTest.php]: Fixed in 3ccaeb0ad8a3ba8f79258d8908ed6bab774a3a41

Comment (liambaloh) [reply tests/MapTest.php]: ...and 1133f756987a045d62dd1194dd36fba0fb5d9ab2 (sorts alphabetically the three added tests)

File: src/Map.php (hunk 3 above)
Comment (liambaloh) [reply src/Map.php]: Good spot! Must have gotten lost in the process. Fixed in c329afdda114015b900470c1ef73edc6439fc1a1

File: README.md [hunk 8]
```diff
@@ -1788,6 +1789,43 @@ Map::from( ['a' => 'X', 'b' => 'Y'] )->has( 'X' );
 ```
 
 
+### if()
+
+Executes callables $then and $else depending on callable $condition.
+
+```php
+public function if( $condition, Callable $then, $Callable $else = null ) : self
```
Comment (aimeos) [inline README.md]: `callable` should be lower case (it's an internal PHP type like "string") and the `$` removed before `$Callable`

File: README.md [hunk 9]
```diff
@@ -1788,6 +1789,43 @@ Map::from( ['a' => 'X', 'b' => 'Y'] )->has( 'X' );
...
+* @param callable|null $else Optional function with (Map) parameter which returns void
+* @return self Same map for fluid interface
+
+**Examples:**
+
+```php
+Map::from( ['a' => 1, 'b' => 0] )->if(
+    fn(Map $map) => $map->has('a'),
+    function(Map $_) {echo "then";},
+    function(Map $_) {echo "else";}
```
Comment (aimeos) [inline README.md]: We add a white space before round brackets like `fn( Map $map ) => $map->has( 'a' ),` and before/after curly braces like `function( Map $_ ) { echo 'then'; },` so it's easier to read.

File: README.md [hunk 10]
```diff
@@ -1788,6 +1789,43 @@ Map::from( ['a' => 'X', 'b' => 'Y'] )->has( 'X' );
...
+);
+//then
+
+Map::from( ['a' => 1, 'b' => 0] )->if(
+    fn(Map $map) => $map->has('c'),
+    function(Map $_) {echo "then";},
+    function(Map $_) {echo "else";}
```
Comment (aimeos) [inline README.md]: Same here

File: README.md [hunk 11]
```diff
@@ -1788,6 +1789,43 @@ Map::from( ['a' => 'X', 'b' => 'Y'] )->has( 'X' );
...
+);
+//else
+
+Map::from( ['a' => 1, 'b' => 0] )->if(
+    fn(Map $map) => $map->has('c'),
+    function(Map $_) {echo "then";}
```
Comment (aimeos) [inline README.md]: And here

File: src/Map.php [hunk 12]
```diff
@@ -3727,4 +3727,32 @@ protected function visit( iterable $entries, array &$result, int $level, ?\Closu
...
+	 * @param callable|null $else Optional function with (Map) parameter which returns void
+	 * @return self Same map for fluid interface
+	 */
+	public function if($condition, callable $then, callable $else = null): self
+	{
+		if($condition instanceof \Closure ? $condition($this) : $condition){
+			$then($this);
+		}elseif($else != null) {
+			$else($this);
+		}
```
Comment (aimeos) [inline src/Map.php]: Please add additional white spaces here too

Comment (liambaloh) [reply src/Map.php]: Resolved in 4e5cf55e0d52299bd0c86f499090473f0c5ab578

File: README.md (hunk 11 above)
Comment (liambaloh) [reply README.md]: Resolved in 4e5cf55e0d52299bd0c86f499090473f0c5ab578

File: README.md (hunk 10 above)
Comment (liambaloh) [reply README.md]: Resolved in 4e5cf55e0d52299bd0c86f499090473f0c5ab578

File: README.md (hunk 9 above)
Comment (liambaloh) [reply README.md]: Resolved in 4e5cf55e0d52299bd0c86f499090473f0c5ab578

File: README.md (hunk 8 above)
Comment (liambaloh) [reply README.md]: Resolved in 4e5cf55e0d52299bd0c86f499090473f0c5ab578

**Comment by liambaloh** [comment]
Requests resolved, however we're now in whitespace territory of code reviews, this is where I usually check out. So if there are additional project-specific style requirements, please apply those yourself as that'll be much faster than explaining them here and having me apply them. Thanks.

File: tests/MapTest.php [hunk 13]
```diff
@@ -1141,6 +1141,36 @@ public function testHasPath()
 	}
 
 
+	public function testIfBlankElse()
+	{
+		Map::from( ['a' => 1, 'b' => 0] )->if(
+			fn(Map $map) => $map->has('c'),
```
Comment (aimeos) [inline tests/MapTest.php]: `fn` is unknown in PHP 7.3 and before. Can you change that to `function` instead?

File: tests/MapTest.php [hunk 14]
```diff
@@ -1141,6 +1141,36 @@ public function testHasPath()
...
+			function(Map $_) { $this->assertTrue(false); }
+		);
+		$this->assertTrue(true);
+	}
+
+
+	public function testIfElse()
+	{
+		Map::from( ['a' => 1, 'b' => 0] )->if(
+			fn(Map $map) => $map->has('c'),
```
Comment (aimeos) [inline tests/MapTest.php]: Same here

File: tests/MapTest.php [hunk 15]
```diff
@@ -1141,6 +1141,36 @@ public function testHasPath()
...
+			function(Map $_) { $this->assertTrue(false); },
+			function(Map $_) { $this->assertTrue(true);}
+		);
+	}
+
+
+	public function testIfThen()
+	{
+		Map::from( ['a' => 1, 'b' => 0] )->if(
+			fn(Map $map) => $map->has('a'),
```
Comment (aimeos) [inline tests/MapTest.php]: And here

File: src/Map.php [hunk 16]
```diff
@@ -3727,4 +3727,32 @@ protected function visit( iterable $entries, array &$result, int $level, ?\Closu
...
+	 *              fn(Map $Map) => $Map->isEmpty(),
+	 *              fn(Map $_) => die("unknown page: " . $pageId))
+	 *          ->first();
+	 *
+	 * @param callable|bool $condition Bool or function with (Map) parameter which evaluates to bool
+	 * @param callable $then Function with (Map) parameter which returns void
+	 * @param callable|null $else Optional function with (Map) parameter which returns void
+	 * @return self Same map for fluid interface
+	 */
+	public function if( $condition, callable $then, callable $else = null ): self
```
Comment (aimeos) [inline src/Map.php]: The second and third arguments can't be a callable like "strcmp", they must be a \Closure instead

File: tests/MapTest.php (hunk 13 above)
Comment (liambaloh) [reply tests/MapTest.php]: resolved in 43ce8d41b8894a4abc84e2be81bd51237264f8ef

File: tests/MapTest.php (hunk 14 above)
Comment (liambaloh) [reply tests/MapTest.php]: resolved in 43ce8d41b8894a4abc84e2be81bd51237264f8ef

File: tests/MapTest.php (hunk 15 above)
Comment (liambaloh) [reply tests/MapTest.php]: resolved in 43ce8d41b8894a4abc84e2be81bd51237264f8ef

File: src/Map.php (hunk 16 above)
Comment (liambaloh) [reply src/Map.php]: resolved in e09020e4f6c9f0232299402636f73131868c0f93

**Comment by aimeos** [comment]
Thanks a lot, now it's good enough for merging :-)

**Comment by liambaloh** [comment]
Great! Thanks for your patience and reviews.
